src/LenderNft.hs

```diff
@@ -30,53 +30,47 @@ import qualified Ledger.Typed.Scripts     as Scripts
 import           Ledger.Value             as Value
 import qualified PlutusTx
 import           PlutusTx.Prelude         hiding (Semigroup (..), unless)
-import qualified Plutus.V1.Ledger.Scripts as Plutus
-import           Prelude                  (Semigroup (..), Show)
 import qualified Common.Utils             as U
 
-{-# INLINABLE lender #-}
-lender :: TokenName
-lender = TokenName { unTokenName = consByteString 76 emptyByteString }  -- L
-
 {-# INLINABLE mkPolicy #-}
-mkPolicy :: ValidatorHash -> TxOutRef -> Integer -> ScriptContext -> Bool
-mkPolicy vh utxo _ ctx = validate
+mkPolicy :: TxOutRef -> ScriptContext -> Bool
+mkPolicy utxo ctx = all validate mintedValue
   where
-    nftIsSentToCollateralSc :: Bool
-    nftIsSentToCollateralSc = traceIfFalse "minted lender nft is not sent to collateral smart contract" (valueOf (U.valueToSc vh ctx) (ownCurrencySymbol ctx) lender == 1)
+    mintFlattened :: [(CurrencySymbol, TokenName, Integer)]
+    mintFlattened = flattenValue $ txInfoMint (scriptContextTxInfo ctx)
+
+    mintedValue :: [(CurrencySymbol, TokenName, Integer)]
+    mintedValue = filter (\(cs, _tn, _n) -> cs == ownCurrencySymbol ctx) mintFlattened
+
+    calculateTokenNameHash :: BuiltinByteString
+    calculateTokenNameHash = sha2_256 (consByteString (txOutRefIdx utxo) ((getTxId . txOutRefId) utxo))
```

Comment:
I think application of `sha2_256` is unnecessary. It's okay to concatenate `txOutRefIdx` and `txOutRefId`, one is a sha256 hash and the other is an integer identifying the output. Concatenation gives you a good enough hash, as good as the sha256 output for `txOutRefId`.

Reply:
`sha2_256` removed

Reply:
@lukipuki I believe concatenation is needed in order for token name byte-string limit.
Following code snippet fails:
`/usr/local/bin/cardano-cli transaction build ...
--mint "1 777e6b4903dab74963ae581d39875c5dac16c09bb1f511c0af1ddda8.4a80a07c2fad5b8fbf9a3f07432ab6e3d31ca4e08077d1484408c1b1a9568f040" --mint-script-file always.succeed ...`
with error:
`AssetName deserisalisation failed: Expected Base16-encoded bytestring, but got "4a80a07c2fad5b8fbf9a3f07432ab6e3d31ca4e08077d1484408c1b1a9568f040"; invalid bytestring size`

Reply:
OK, we can then add back `sha2_256`, it's actually a fairly cheap operation in the Plutus cost model.